### PR TITLE
`vcov` version 2

### PR DESCRIPTION
**Pull request by vincentarelbundock**

### CODE REVIEW

File: R/1_model_parameters.R
Comment (bwiernik) [inline R/1_model_parameters.R:320]:
```suggestion
#'   - Robust estimation (i.e., `vcov` set to a value other than `NULL`) of standardized parameters only
```

Should refer to the correct argument

File: R/1_model_parameters.R
Comment (bwiernik) [inline R/1_model_parameters.R:497]: Should we add a check for `robust` in the dots and give a warning too?

File: R/1_model_parameters.R
Comment (bwiernik) [inline R/1_model_parameters.R:544]:
```suggestion
      vcov_args = vcov_args
    )
```

File: R/1_model_parameters.R
Comment (bwiernik) [inline R/1_model_parameters.R:624]:
```suggestion
    vcov_args = vcov_args
  )
```

File: R/3_p_value.R
Comment (bwiernik) [inline R/3_p_value.R:13]:
```suggestion
#' @param method For linear mixed models, `method` can be
```

File: R/3_p_value.R
Comment (bwiernik) [inline R/3_p_value.R:61]: Do we want to give a depreciation warning?

File: R/4_standard_error.R
Comment (bwiernik) [inline R/4_standard_error.R:12]:
```suggestion
#' @param vcov Variance-covariance matrix used to compute uncertainty estimates (e.g., for robust standard errors). This argument accepts a covariance matrix, a function which returns a covariance matrix, or a string which identifies the function to be used to compute the covariance matrix.
```

File: R/4_standard_error.R
Comment (bwiernik) [inline R/4_standard_error.R:84]: Do we want to give a deprecation warning?

File: R/4_standard_error.R
Comment (bwiernik) [inline R/4_standard_error.R:180]: Do we want a warning?

File: R/4_standard_error.R
Comment (bwiernik) [inline R/4_standard_error.R:173]: Do we want a warning?

File: R/4_standard_error.R
Comment (bwiernik) [inline R/4_standard_error.R:192]: Let's use switch() here instead eg,

```suggestion
  switch(vcov_fun,  HC0 =, HC1 =, HC2 =, HC3 =, HC4 =, HC5 = ...
```

File: R/4_standard_error.R
Comment (vincentarelbundock) [reply R/4_standard_error.R:173]: I will include a warning in `.get_vcov()` since everything flows through there.

File: R/4_standard_error.R
Comment (vincentarelbundock) [reply R/4_standard_error.R:180]: I will include a warning in `.get_vcov()` since everything flows through there.

File: R/4_standard_error.R
Comment (vincentarelbundock) [reply R/4_standard_error.R:84]: I will include a warning in `.get_vcov()` since everything flows through there.

File: R/3_p_value.R
Comment (vincentarelbundock) [reply R/3_p_value.R:61]: I will include a warning in `.get_vcov()` since everything flows through there.

**Comment by vincentarelbundock** [comment]
@bwiernik I'm very confused now. Looks like we tried to push at the same time. I didn't know you could push to a branch of my fork...

Sorry if there's a conflict now. I'll close the computer for a while. Let me know if/when I can pick it back up.

**Comment by bwiernik** [comment]
Default for PRs in these repos is that maintainers can add commits. I'm done. Your force push resolved conflicts. Sorry for the confusion

File: R/4_standard_error.R
Comment (bwiernik) [inline R/4_standard_error.R:8]: Not related to your changes, but I don't understand what this argument is supposed to do, and it's not used anywhere. Should we remove it?

Comment (strengejacke) [reply R/4_standard_error.R:8]: https://github.com/easystats/parameters/blob/f957b8f71444ca24da0762ae583ad8c0d5e75fce/R/methods_base.R

Comment (bwiernik) [reply R/4_standard_error.R:8]: Oh, okay. I didn't realize that these functions could be applied to vectors to compute the standard error of the mean and not only models. 

We should clarify that for the `model` and `force` arguments. 

```
model A model. Can also be a vector or data frame to compute the standard error of the mean for one or more variables. 
force When `model` is a vector or data frame, if `force = TRUE`, ...
```

Comment (vincentarelbundock) [reply R/4_standard_error.R:8]: Ooof. Is this `force` argument ever a good idea? Feels like users should be responsible for doing dumb stuff on their own... (Also, the word is pretty generic; could lead to confusion.)

Comment (strengejacke) [reply R/4_standard_error.R:8]: It's kind of "historical reason", we use the `force` argument in different places (related to factors), e.g. in `standardize()` to force standardization of factors as well.

>  Feels like users should be responsible for doing dumb stuff

With this attitude, we'll never come beyond "p < 0.05 means evidence of an effect" ;-) We have an educational function / responsibility as well.

Comment (strengejacke) [reply R/4_standard_error.R:8]: toni

(this smiley was named after me, because every now and then I appear as wisenheimer :-))

**Comment by vincentarelbundock** [comment]
Does anyone know why `pkgdown` is failing here? The site builds fine and passes all checks on my machine. Also, the `predict.parameters_clusters()` roxygen seems pretty sandard. The error message: 

```
2022-02-22T15:00:12.5208446Z Error in `check_missing_topics()`:
2022-02-22T15:00:12.5209451Z ! Topics missing from index: predict.parameters_clusters
2022-02-22T15:00:12.5210036Z Backtrace:
2022-02-22T15:00:12.5212750Z      ▆
2022-02-22T15:00:12.5213945Z   1. └─pkgdown::build_site_github_pages(new_process = FALSE, install = FALSE)
2022-02-22T15:00:12.5214669Z   2.   └─pkgdown::build_site(...)
2022-02-22T15:00:12.5215346Z   3.     └─pkgdown:::build_site_local(...)
2022-02-22T15:00:12.5216552Z   4.       └─pkgdown::build_reference(...)
2022-02-22T15:00:12.5217524Z   5.         └─pkgdown::build_reference_index(pkg)
2022-02-22T15:00:12.5218693Z   6.           ├─pkgdown::render_page(...)
2022-02-22T15:00:12.5219577Z   7.           │ └─pkgdown:::render_page_html(pkg, name = name, data = data, depth = depth)
2022-02-22T15:00:12.5220347Z   8.           │   └─utils::modifyList(data_template(pkg, depth = depth), data)
2022-02-22T15:00:12.5221595Z   9.           │     └─base::stopifnot(is.list(x), is.list(val))
2022-02-22T15:00:12.5222132Z  10.           └─pkgdown:::data_reference_index(pkg)
2022-02-22T15:00:12.5222632Z  11.             └─pkgdown:::check_missing_topics(rows, pkg)
2022-02-22T15:00:12.5223088Z  12.               └─rlang::abort(text)
2022-02-22T15:00:12.5223356Z Execution halted
2022-02-22T15:00:12.5410475Z ##[error]Process completed with exit code 1.
```

**Comment by strengejacke** [comment]
I guess

```
Error in `check_missing_topics()`:
2022-02-22T15:00:12.5209451Z ! Topics missing from index: predict.parameters_clusters
```

**Comment by vincentarelbundock** [comment]
> I guess
> 
> ```
> Error in `check_missing_topics()`:
> 2022-02-22T15:00:12.5209451Z ! Topics missing from index: predict.parameters_clusters
> ```

Yes, of course I can see that. But the `.Rd` file seems fine, and the site index builds on my local machine without issues...

**Comment by strengejacke** [comment]
Ah, ok, got it. Maybe temporary, or something we can ignore?

**Comment by vincentarelbundock** [comment]
Could be. It worries me a bit, but I guess we could merge and I can commit to fixing it later if it still breaks on main.

The other failure is a pandoc install fail on Windows, so it's not related to the PR.

Aside from that, I think this is good to go, so I removed the WIP label in the title.

Let me know what you think.

PS: I commit to fixing any issues related to this as quickly as possible if/when they come up.

File: R/ci_generic.R
Comment (strengejacke) [inline R/ci_generic.R:105]: you could omit `se_kenward(model)` here, `switch()` allows catching multiple values here.

**Comment by strengejacke** [comment]
It's hard to review all the changes. I would suggest adding short tests for robust estimation that work for the current main-branch, and then merge those tests into this PR and see if these tests also pass for this PR. There were several improvements/bug fixes related to robust estimation in the past and I think we have a pretty "robust" (haha) code base that works correctly. I don't want to introduce any new issues that "revert" some of the past bug fixes ;-)

**Comment by mattansb** [comment]
image

File: tests/testthat/test-model_parameters_robust.R
Comment (strengejacke) [inline tests/testthat/test-model_parameters_robust.R:54]: @vincentarelbundock here we should add the tests for this PR. Results _should_ be identical. We could also add tests for `ci_ml1()` and similar for robust SE.

**Comment by vincentarelbundock** [comment]
@strengejacke I kept all of the old tests in *exactly* as they were, including the deprecated arguments. The only change I made to them is adding `verbose=FALSE` to some to silence the new warning.

For example,

[Here we have:](https://github.com/vincentarelbundock/parameters/blob/simple_vcov2/tests/testthat/test-model_parameters_robust.R#L18)

* `ci(robust = TRUE)`
* `model_parameters(mod, robust = TRUE)`
* `model_parameters(mod, vcov_estimation = "CR", vcov_type = "CR1")`
* `model_parameters(mod, standardize = "refit", robust = TRUE)`
* `model_parameters(vcov_args)`

[Here we have:](https://github.com/vincentarelbundock/parameters/blob/simple_vcov2/tests/testthat/test-model_parameters_std_mixed.R#L82)

* `model_parameters(vcov_estimation, vcov_type, vcov_args)` for `merMod` models.
* `model_parameters(robust = TRUE)`

Everything works exactly as before for these arguments. In addition, I added new tests for the new behavior to see if it matched the old ones.

To clarify, are you requesting new tests for previously untested behavior?

**Comment by vincentarelbundock** [comment]
@strengejacke did you revert the `test-model_parameters_robust.R` file?!

It included all of the original old tests plus a bunch of new tests where I compared new output to old one. Now we lost all my new added tests!

File: tests/testthat/test-model_parameters_robust.R
Comment (vincentarelbundock) [reply tests/testthat/test-model_parameters_robust.R]: I had this already, but it looks like you reverted the file.

Comment (vincentarelbundock) [reply tests/testthat/test-model_parameters_robust.R]: I did not have tests for `ci_ml1` but I don't know this function.

**Comment by strengejacke** [comment]
> @strengejacke did you revert the `test-model_parameters_robust.R` file?!
> 
> It included all of the original old tests plus a bunch of new tests where I compared new output to old one. Now we lost all my new added tests!

Not sure, I was trying to resolve some conflicts... 😬
So we should add back the tests removed here: https://github.com/easystats/parameters/pull/661/commits/cfe002bef6e3574c3ca4f85fa64258e187d744a5

?

**Comment by strengejacke** [comment]
Mainly, I wanted to make sure that the new changes replicate previous test results. But I then saw that we indeed haven't had tests for all use cases... And I encountered an issue with `ci_ml1()` and the `robust` option for the current main branch

**Comment by vincentarelbundock** [comment]
> > @strengejacke did you revert the `test-model_parameters_robust.R` file?!
> > It included all of the original old tests plus a bunch of new tests where I compared new output to old one. Now we lost all my new added tests!
> 
> Not sure, I was trying to resolve some conflicts... 😬 So we should add back the tests removed here: [cfe002b](https://github.com/easystats/parameters/commit/cfe002bef6e3574c3ca4f85fa64258e187d744a5)
> 
> ?

Yep, I think that's the commit that overwrote the tests. I'm not git-confident enough to revert that. Can you do it?

I can look into `ml1` eventually.

**Comment by vincentarelbundock** [comment]
Not 100% sure, but it looks like some tests were also overwritten in the previous batch of merge.

Next time it's probably best to merge only at the end before we've "agreed" on a final merge.

**Comment by vincentarelbundock** [comment]
Also more "surgical" with the scope of my pull requests ;)

**Comment by strengejacke** [comment]
I think there have been no changes/commits to test-model_parameters_robust.R from you...  Or is it because I'm watching this page from my mobile phone so I miss something?

**Comment by vincentarelbundock** [comment]
I have definitely made changes to this file. For example, I had changed the file to use `requiet` to skip tests when packages are unavailable instead of using a big loop. Here you can see that "my style" is overwritten with the old style:

https://github.com/easystats/parameters/commit/cfe002bef6e3574c3ca4f85fa64258e187d744a5#diff-40b26649e4d99f6ecc4f20b80a0a5714e17ab831abab7827a2dbb1476cb5aec9

I'm not sure if anything else was overwritten, but it's probably not that much.

**Comment by strengejacke** [comment]
> I'm not sure if anything else was overwritten, but it's probably not that much.

Since I update this PR from main, this test should be the only one that got overwritten, but only due to my "resolving conflicts" activity.

File: R/methods_kmeans.R
Comment (strengejacke) [inline R/methods_kmeans.R:126]: I think this is the reason why pkgdown fails, because `predict.parameters_clusters()` is now indexed in package-down. You need to add an entry `_pkgdown.yaml` as well, if you create a help page for this method.

**Review by strengejacke — CHANGES_REQUESTED** [review]

**Comment by strengejacke** [comment]
> Next time it's probably best to merge only at the end before we've "agreed" on a final merge.

Regarding the many files that have been changed, I think it's not bad to sync master/main and PRs, so you can safely make commits to both master/main and PR branches, w/o the risk of ending up in "unresolvable" conflicts. That's why I update PRs every now and then from main/master. When we make changes to file A in main, and this PR requires changes in file A as well at a later stage, we ensured that file A is in sync, so no conflicts arise.

File: _pkgdown.yml
Comment (vincentarelbundock) [inline _pkgdown.yml:102]: Ha! didn't know about this. Thanks!

**Comment by vincentarelbundock** [comment]
> > Next time it's probably best to merge only at the end before we've "agreed" on a final merge.
> 
> Regarding the many files that have been changed, I think it's not bad to sync master/main and PRs, so you can safely make commits to both master/main and PR branches, w/o the risk of ending up in "unresolvable" conflicts. That's why I update PRs every now and then from main/master. When we make changes to file A in main, and this PR requires changes in file A as well at a later stage, we ensured that file A is in sync, so no conflicts arise.

Yeah, that makes sense. I personally prefer to do all the work then rebase on main. Either way is fine generally. I just got confused because I didn't expect others to make commits in my branch.

I think next time I'll make an official PR later in the process and keep it in a private branch for longer to avoid confusion and needless review.

Do you want to wait for others before merge.

**Comment by strengejacke** [comment]
Before we do the final step (merging), we call fireman @DominiqueMakowski to update the news-file...  😁

**Comment by strengejacke** [comment]
Looks like everything is ok now. Maybe we can add some more tests later, specifically for mixed models and the combination of ci-methods and robust (i.e. `ci_ml1()` or `ci_satterthwaite()` with robust SE). And, if I'm not mistaken, there are no robust SE for kenward-rogers DF, right? Because KR computes its own vcov?

**Comment by strengejacke** [comment]
And @vincentarelbundock , you could now move on to https://github.com/easystats/insight/issues/505 :-)

**Comment by vincentarelbundock** [comment]
> And @vincentarelbundock , you could now move on to [easystats/insight#505](https://github.com/easystats/insight/issues/505) :-)

Will do. But work is a bit crazy now, so it's going to take me a little bit.

**Comment by strengejacke** [comment]
> Will do. But work is a bit crazy now, so it's going to take me a little bit.

There's no hurry. You can also do that next week.

**Comment by bwiernik** [comment]
@vincentarelbundock Do you think you might be able to incorporate this framework into get_predicted()/estimate_*() ?

**Comment by strengejacke** [comment]
Yes, I think this was his intention... https://github.com/easystats/insight/issues/505

**Comment by bwiernik** [comment]
Ah didn't see that!

**Pull request merged**
